mne/stats/cluster_level.py

```diff
@@ -1729,3 +1742,465 @@ def summarize_clusters_stc(
     data_summary[:, 0] = np.sum(data_summary, axis=1)
 
     return klass(data_summary, vertices, tmin, tstep, subject)
+
+
+def validate_input_dataframe(df: pd.DataFrame, formula: str):
+    """
+    Validate the input dataframe for the cluster permutation test.
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        Dataframe with 3 columns (subject_index, condition, data).
+    formula : formulaic.ModelSpec
+        Wilkinson style Formula for the design matrix.
+
+    Returns
+    -------
+    dv_name : str
+        Name of the dependent variable.
+    """
+    # extract dependent variable name from formula
+    formulaic = _soft_import(
+        "formulaic", purpose="set up Design Matrix"
+    )  # soft import (not a dependency for MNE)
+    formula = formulaic.Formula(formula)
+    dv_name = str(formula.lhs)
+
+    # check if all necessary columns are present
+    if dv_name not in df.columns:
+        raise ValueError("""DataFrame needs to contain a column
+                        with the dependent variable name
+                        as defined in the formula""")
+    if "condition" not in df.columns:
+        raise ValueError("DataFrame needs to contain a condition column")
+    if "subject_index" not in df.columns:
+        raise ValueError("DataFrame needs to contain a subject_index column")
+
+    # check if the data column contains only valid types
+    check_column_types(df[dv_name])
+
+    # check if the shape of the data is consistent
+    if not all(data.data.shape == df[dv_name][0].data.shape for data in df[dv_name]):
+        raise ValueError("Data objects need to have the same shape")
+
+    # check if the condition column contains only 2 unique values
+    if len(pd.unique(df.condition)) != 2:
+        raise ValueError("currently only supports 2 conditions.")
+
+    return dv_name
+
+
+def check_column_types(input_data: np.ndarray):
```

Comment:
I believe you could annotate this with np.typing.NDArray[np.float_] but i'm not sure if it's actually relevant here